File: 2019/day9/day9.py
#!/usr/bin/env python3

import itertools
from collections import deque


FEEDBACK_LOOP_ENABLED = False

num_of_parameters = {
    1: 3,
    2: 3,
    3: 1,
    4: 1,
    5: 2,
    6: 2,
    7: 3,
    8: 3,
    9: 1
}


class Amplifier:

    def __init__(self, data, inputs=None):
        self.data = data.copy()
        self._output = deque()
        self._input = deque()
        self.instruction_pointer = 0
        self.end = 0
        if inputs is not None:
            self.input = inputs
        self._next = Amplifier
        self.completed = False
        self.relative_base = 0
        self.num_of_instructions = 0

    @property
    def input(self):
        return self._input.popleft()

    @input.setter
    def input(self, v):
        try:
            self._input.extend(v)
        except TypeError:
            self._input.append(v)

    @property
    def output(self):
        return self._output.popleft()

    @output.setter
    def output(self, v):
        self._output.append(v)

    def run(self):
        while True:
            start = self.instruction_pointer
            opcode, param_modes = parse_opcode(self.data[start])
            if opcode == 99:
                print('this progrom has ended')
                self.completed = True
                return

            self.end = start + num_of_parameters[opcode] + 1
            parameters = self.data[start + 1:self.end]
            params = list(itertools.zip_longest(param_modes,
                                                parameters,
                                                fillvalue=0))

            self.compute(opcode, params)
            self.num_of_instructions += 1
            if len(self._output) > 0 and FEEDBACK_LOOP_ENABLED:
                return

    def compute(self, opcode, params):
        def parameter_value(mode_and_param, array):
            mode, param = mode_and_param
            if mode == 0:
                return self.get_data(param)
            elif mode == 1:
                return param
            elif mode == 2:
                return self.get_data(self.relative_base + param)

        # These opcodes compare 2 params and write to the location of the 3rd
        if opcode in [1, 2, 7, 8]:
            p1 = parameter_value(params[0], self.data)
            p2 = parameter_value(params[1], self.data)
            writer = params[-1]
            value = 0
            if opcode == 1:
                value = p1 + p2
            elif opcode == 2:
                value = p1 * p2
            elif opcode == 7:
                if p1 < p2:
                    value = 1
            elif opcode == 8:
                if p1 == p2:
                    value = 1
            if value is None:
                print(self.instruction_pointer)
                raise

            self.write_data(writer, value)

        # These opcodes reassign the terminal pointer
        elif opcode in [5, 6]:
            p1 = parameter_value(params[0], self.data)
            p2 = parameter_value(params[1], self.data)
            if opcode == 5:
                if p1 != 0:
                    self.instruction_pointer = p2
                    return
            elif opcode == 6:
                if p1 == 0:
                    self.instruction_pointer = p2
                    return

        # These opcodes only take one parameter
        elif opcode in [3, 4, 9]:
            p1 = params[0]
            if opcode == 3:
                self.write_data(p1, self.input)
            elif opcode == 4:
                self.output = parameter_value(p1, self.data)
            elif opcode == 9:
                self.relative_base += parameter_value(p1, self.data)

        self.instruction_pointer = self.end

    def reset(self):
        self.instruction_pointer = 0

    def write_data(self, k, v):
        mode, pos = k
        if mode == 2:
            pos += self.relative_base
        try:
            self.data[pos] = v
        except IndexError:
            self._grow(pos + 1 - len(self.data))
            self.data[pos] = v

    def get_data(self, k):
        try:
            return self.data[k]
        except IndexError:
            return 0

    def _grow(self, size):
        for x in range(size + 1):
            self.data.append(0)


def parse_opcode(opcode):
    if opcode < 10:
        return (opcode, [])
    else:
        op = opcode % 100
        params = opcode // 100
        pos_nums = []
        while params != 0:
            pos_nums.append(params % 10)
            params //= 10
        return (op, pos_nums)


def phase_settings(low, high):
    for x in itertools.permutations(range(low, high)):
        yield x


def part1(intcode):
    amp = Amplifier(intcode, 1)
    amp.run()
    return amp.output


def part2(intcode):
    amp = Amplifier(intcode, 2)
    amp.run()
    return amp.output


if __name__ == '__main__':
    with open('input.txt') as f:
        data = f.read().split(',')
    intcode = [int(x) for x in data]
    print(part1(intcode))
    print(part2(intcode))

File: 2019/day9/day9.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

type parameter struct {
	mode  int
	value int
}

func newParam(mode, value int) parameter {
	return parameter{
		mode:  mode,
		value: value,
	}
}

type program struct {
	data    []int
	rb      int
	ip      int
	inputs  chan int
	outputs chan int
}

func new(data []int) program {
	p := program{
		data:    data,
		rb:      0,
		ip:      0,
		inputs:  make(chan int, 2),
		outputs: make(chan int, 2),
	}
	return p
}

func (p *program) run() {
	for {
		opcode, modes := parseOpcode(p.data[p.ip])
		if opcode == 99 {
			fmt.Println("This program has ended")
			return
		}
		start := p.ip + 1
		end := start + numOfParams(opcode)
		rawParams := p.data[start:end]
		params := makeParams(rawParams, modes)
		p.compute(opcode, params)
		if p.ip == start-1 {
			p.ip = end
		}
	}
}

func (p *program) compute(opcode int, params []parameter) {
	switch opcode {
	case 1, 2, 7, 8:
		p1 := p.get(params[0])
		p2 := p.get(params[1])
		writer := params[2]
		var value int
		switch opcode {
		case 1:
			value = p1 + p2
		case 2:
			value = p1 * p2
		case 7:
			if p1 < p2 {
				value = 1
			}
		case 8:
			if p1 == p2 {
				value = 1
			}
		}
		p.set(writer, value)

	case 5, 6:
		p1 := p.get(params[0])
		p2 := p.get(params[1])
		switch opcode {
		case 5:
			if p1 != 0 {
				p.ip = p2
				return
			}
		case 6:
			if p1 == 0 {
				p.ip = p2
				return
			}
		}

	case 3, 4, 9:
		p1 := params[0]
		switch opcode {
		case 3:
			i := <-p.inputs
			p.set(p1, i)
		case 4:
			p.outputs <- p.get(p1)
		case 9:
			p.rb += p.get(p1)
		}
	}

}

func (p *program) addInput(i int) {
	if p.inputs == nil {
		p.inputs = make(chan int, 2)
	}
	p.inputs <- i
}

func (p *program) get(param parameter) int {
	switch param.mode {
	case 0:
		return p.data[param.value]
	case 1:
		return param.value
	case 2:
		return p.data[p.rb+param.value]
	}
	return 0
}

func (p *program) set(w parameter, value int) {
	if w.mode == 2 {
		w.value += p.rb
	}
	if w.value >= len(p.data) {
		p.grow(w.value + 1 - len(p.data))
	}
	p.data[w.value] = value
}

func (p *program) grow(size int) {
	d := make([]int, len(p.data)+size)
	for i := 0; i < len(p.data); i++ {
		d[i] = p.data[i]
	}
	p.data = d
}

func numOfParams(opcode int) int {
	switch opcode {
	case 1, 2, 7, 8:
		return 3
	case 3, 4, 9:
		return 1
	case 5, 6:
		return 2
	}
	return 0
}

func setup() []int {
	data, err := ioutil.ReadFile("input.txt")
	if err != nil {
		panic(err)
	}
	intcode := make([]int, 0, len(data))
	for _, s := range strings.Split(string(data), ",") {
		number, err := strconv.Atoi(s)
		if err != nil {
			fmt.Printf("%s is not an integer", s)
			continue
		}
		intcode = append(intcode, number)
	}
	return intcode
}

func parseOpcode(op int) (int, []int) {
	var modes []int
	if op < 100 {
		return op, modes
	}
	o := op % 100
	m := op / 100
	for m != 0 {
		modes = append(modes, m%10)
		m /= 10
	}
	return o, modes
}

func makeParams(params, modes []int) []parameter {
	pm := make([]parameter, 0, len(params))
	for i, value := range params {
		mode := 0
		if i < len(modes) {
			mode = modes[i]
		}
		pm = append(pm, newParam(mode, value))
	}
	return pm
}

func part1(intcode []int) {
	p := new(intcode)
	p.addInput(1)
	p.run()
	fmt.Println(<-p.outputs)
}

func part2(intcode []int) {
	p := new(intcode)
	p.addInput(2)
	p.run()
	fmt.Println(<-p.outputs)
}

func main() {
	intcode := setup()
	part1(intcode)
	part2(intcode)
}
